go/vrf/ecvrf_p256_sha256_swu.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vrf
+
+import (
+	"crypto"
+	"crypto/elliptic"
+	"math/big"
+)
+
+type (
+	p256SHA256SWUSuite struct{ *ECVRFParams }
+	p256SHA256SWUAux   struct {
+		p256SHA256TAIAux
+		params *ECVRFParams
+	}
+)
+
+var p256SHA256SWU p256SHA256SWUSuite
+
+func initP256SHA256SWU() {
+	// https://tools.ietf.org/html/draft-irtf-cfrg-vrf-06#section-5.5
+	// This ciphersuite is identical to ECVRF-P256-SHA256-TAI except that
+	// the ECVRF_hash_to_curve function and suite_string = 0x02
+	p := &ECVRFParams{
+		suite:    0x02,            // int_to_string(1, 1)
+		ec:       elliptic.P256(), // NIST P-256 elliptic curve, [FIPS-186-4] (Section D.1.2.3).
+		fieldLen: 32,              // Params().BitSize / 8 = 2n. Must be a multiple of 2.
+		qLen:     32,              // Params().N.BitLen
+		ptLen:    33,              // Size of encoded EC point
+		cofactor: big.NewInt(1),
+		hash:     crypto.SHA256,
+	}
+	p.aux = p256SHA256SWUAux{
+		p256SHA256TAIAux: p256SHA256TAIAux{params: p},
+		params:           p,
+	}
+	p256SHA256SWU.ECVRFParams = p
+}
+
+// Params returns the parameters for the ECVRF.
+func (s p256SHA256SWUSuite) Params() *ECVRFParams { return s.ECVRFParams }
+
+// HashToCurve implements the Simplified SWU algorithm from section 5.4.1.3.
+// SWU is implemented with running time that is independent of the input
+// alpha (so-called "constant-time").
```

Comment:
There should probably be some big warning here, as the big.int operations are not necessarily constant-time in Go (https://github.com/golang/go/issues/20654, https://github.com/binance-chain/tss-lib/issues/44). This probably also affects other parts of the implementation, and we should check if there is any risk.

Reply:
Added the warning.  Will need to think about the constant time bit for the other parts of the impl.